pkg/tun/tun.go

```diff
@@ -0,0 +1,164 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// Package tun wraps the wg tunnel implementation and adds packet filters.
+package tun
+
+import (
+	"encoding/binary"
+	"fmt"
+	"net/netip"
+	"os"
+
+	"golang.zx2c4.com/wireguard/tun"
+)
+
+// PacketHeader is the decoded representation of the wireguard packet.
+type PacketHeader struct {
+	SourceAddr      netip.Addr
+	DestinationAddr netip.Addr
+	FlowLabel       uint32
+	PayloadLength   uint16
+	Version         uint8
+	TrafficClass    uint8
+	NextHeader      uint8
+	HopLimit        uint8
+}
+
+// PacketFilter is the function for filtering the packets going through the wireguard tunnel.
+type PacketFilter func(PacketHeader) bool
+
+// FilterAllExceptIP drops all packets that have the destination different from the ip specified.
+func FilterAllExceptIP(ip netip.Addr) PacketFilter {
+	return func(p PacketHeader) bool {
+		return p.DestinationAddr != ip
+	}
+}
+
+// Device wraps wg tunnel device.
+type Device struct {
+	Tun           tun.Device
+	PacketFilters []PacketFilter
+}
+
+// CreateTUN creates a Device with the provided name and MTU.
+func CreateTUN(iface string, mtu int, packetFilters ...PacketFilter) (*Device, error) {
+	tun, err := tun.CreateTUN(iface, mtu)
+	if err != nil {
+		return nil, fmt.Errorf("error creating tun device: %w", err)
+	}
+
+	return &Device{Tun: tun, PacketFilters: packetFilters}, nil
+}
+
+// File returns the file descriptor of the device.
+func (d *Device) File() *os.File {
+	return d.Tun.File()
+}
+
+// Read one or more packets from the Device (without any additional headers).
+// On a successful read it returns the number of packets read, and sets
+// packet lengths within the sizes slice. len(sizes) must be >= len(bufs).
+// A nonzero offset can be used to instruct the Device on where to begin
+// reading into each element of the bufs slice.
+func (d *Device) Read(bufs [][]byte, sizes []int, offset int) (int, error) {
+	return d.Tun.Read(bufs, sizes, offset)
+}
+
+// Write one or more packets to the device (without any additional headers).
+// On a successful write it returns the number of packets written. A nonzero
+// offset can be used to instruct the Device on where to begin writing from
+// each packet contained within the bufs slice.
+func (d *Device) Write(bufs [][]byte, offset int) (int, error) {
+	if len(d.PacketFilters) > 0 {
+		result := bufs[:0]
+
+		for _, buf := range bufs {
+			packet, err := d.decodePacketHeader(buf)
+			if err != nil {
+				return 0, err
```

Comment:
we should drop a single invalid packet, not all of them

Reply:
I have another idea, let me change that

Reply:
updated to always drop the packets we can't decode for any reason